src/v/pandaproxy/rest/proxy.cc

```diff
@@ -9,52 +9,115 @@
 
 #include "pandaproxy/rest/proxy.h"
 
+#include "config/property.h"
+#include "config/rest_authn_endpoint.h"
+#include "net/unresolved_address.h"
 #include "pandaproxy/api/api-doc/rest.json.h"
 #include "pandaproxy/logger.h"
+#include "pandaproxy/parsing/exceptions.h"
+#include "pandaproxy/parsing/from_chars.h"
 #include "pandaproxy/rest/configuration.h"
 #include "pandaproxy/rest/handlers.h"
+#include "pandaproxy/types.h"
+#include "redpanda/request_auth.h"
+#include "utils/gate_guard.h"
 
 #include <seastar/core/future-util.hh>
 #include <seastar/core/memory.hh>
+#include <seastar/core/temporary_buffer.hh>
 #include <seastar/http/api_docs.hh>
+#include <seastar/http/httpd.hh>
+#include <seastar/http/request_parser.hh>
+
+#include <algorithm>
+#include <string>
 
 namespace pandaproxy::rest {
 
+namespace {
+net::unresolved_address address_from_host(ss::sstring host_and_port) {
+    auto colon = host_and_port.find_last_of(':');
+    ss::sstring host{host_and_port.substr(0, colon)};
+    ss::sstring port_str{host_and_port.substr(colon+1)};
+
+    auto res = parse::from_chars<uint16_t>{}(port_str);
+    if (res.has_error()) {
+        throw parse::error(
+          parse::error_code::not_acceptable, "INT conversion fail");
+    }
+
+    return net::unresolved_address{host, res.value()};
+}
+} // namespace
+
 using server = ctx_server<proxy>;
 
-server::routes_t get_proxy_routes() {
+template<typename Handler>
+auto wrap(configuration& cfg, Handler h) {
+    return [&cfg, h{std::move(h)}](
+             server::request_t rq,
+             server::reply_t rp) -> ss::future<server::reply_t> {
+        net::unresolved_address req_address{address_from_host(rq.req->get_header("Host"))};
+        rq.ctx.authn_type = rq.service().config().authn_type(req_address);
```

Comment:
This is the commit where we add the `authn_type()` method to the configuration file.